### PR TITLE
Added docker-compose support for quick GDPS testing setup

### PR DESCRIPTION
**Pull request by cn-ml**
Hello @Cvolton,

today I tried setting up my own GPDS using your software and I found it was not as easy as I initially expected, even though I would call myself relatively good with software. So here I added a docker-compose.yml file for quick setup.

To create a new GDPS on your local machine just run `docker-compose up` in the project root and then you're ready to go. The database is automatically created and initialized from the repository (currently not persistent, but that can be added as a one-liner in the compose file). Instead of replacing every link in the GeometryDash.exe file I just modified my host file to point to my local machine.

I also pre-configured the database connection to point to the docker container of the database with the preset username.

There is no contribution file in the repository so I didn't know if you accept pull requests, but for me it was good to have docker support so here it is, do with it what you want.

OnlyFreeUsername

### CODE REVIEW

**Comment by Cvolton** [comment]
Thanks for this. I'm somewhat worried about updates though (which TBF aren't easy with a manual setup either).  As I'm not all too familiar with Docker, is there any mechanism to facilitate them in an easy manner, accounting for changes to both the PHP files and the SQL structure?

**Comment by OnlyFreeUsername** [comment]
What exactly are you worried about? Are you worried about the persistence of the database?
Right now there are two containers in the composer stack. First being the php apache file server that just serves the php files (if I understood php correctly (which I'm not particularly familiar with) there is no data persisting on this container, just the plain php files). The connection to the database ("db" container) connects the file server to the database that contains the MySQL database. If you want to update the PHP files the database container is untouched.

The other thing is the persistence of the database itself. By default the database with perish once the container is destroyed. I can add a bind or volume mount to the database container to persist the database on the host filesystem.

Can you verify that the all data that needs to persist between updates is contained in the mysql database?

**Comment by OnlyFreeUsername** [comment]
Okay, I just noticed that you save the levels as files on the php server and deliver from the filesystem directly:

https://github.com/Cvolton/GMDprivateServer/blob/51c6fb4ab6e3fe7584c4d79f15056a489c7a16ad/incl/levels/downloadGJLevel.php#L82-L83

Why did you choose to exclude the levelString from the database:

https://github.com/Cvolton/GMDprivateServer/blob/51c6fb4ab6e3fe7584c4d79f15056a489c7a16ad/incl/levels/uploadGJLevel.php#L145

**Comment by OnlyFreeUsername** [comment]
Now I did a little workaround and added a volume to the php server to persist the levels and backups. that seems to be working now.

**Comment by Jecket22** [comment]
> Why did you choose to exclude the levelString from the database? @Cvolton

I think at some point it was considered that saving the levels on disk would be better than saving it on the database due to the sizes of levels (although I'm not too sure if this is a huge concern or a performance issue)

(also sorry for the notification bump, I just saw this pull request and this looks pretty interesting)

File: config/connection.php [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?php
-$servername = "127.0.0.1";
+$servername = "db";
```
Comment (Jecket22) [inline config/connection.php:2]: Also I suggest not changing this (unless this is specific to docker? This is just the address the mysql connection should listen to so)

Comment (NayLex) [reply config/connection.php:2]: Docker-specific thing. It should be like this.

**Comment by Cvolton** [comment]
> > Why did you choose to exclude the levelString from the database? @Cvolton
> 
> I think at some point it was considered that saving the levels on disk would be better than saving it on the database due to the sizes of levels (although I'm not too sure if this is a huge concern or a performance issue)
> 
> (also sorry for the notification bump, I just saw this pull request and this looks pretty interesting)

this was definitely a performance issue back when indexes weren't set up properly (remember that levelStrings can go up to 50 MB in size too), not sure if it still applies nowadays.

My earlier comment about updates stemmed from the fact that I missed that the note that the container wasn't persistent, so I was worried about SQL structure updates. Though realistically, in the worst case scenario, that can be handled in the exact same way it's handled with manual setups (which is quite frankly not very good and an issue I will have to work on too).

I will have to look into how this is usually handled, as Dockerfiles are generally kept separate from the main repos of the projects they work with

**Pull request closed without merging**